Sources/Testing/ExitTests/WaitFor.swift

```diff
@@ -12,36 +12,123 @@
+  /// `siginfo_t` type.
+  ///
+  /// - Parameters:
+  ///   - siginfo: The instance of `siginfo_t` to initialize from.
+  ///
+  /// - Throws: If `siginfo.si_code` does not equal either `CLD_EXITED`,
+  ///   `CLD_KILLED`, or `CLD_DUMPED` (i.e. it does not represent an exit
+  ///   condition.)
+  fileprivate init(_ siginfo: siginfo_t) throws {
+    switch siginfo.si_code {
+    case .init(CLD_EXITED):
+      self = .exitCode(siginfo.si_status)
+    case .init(CLD_KILLED), .init(CLD_DUMPED):
+      self = .signal(siginfo.si_status)
+    default:
+      throw SystemError(description: "Unexpected siginfo_t value. Please file a bug report at https://github.com/apple/swift-testing/issues/new and include this information: \(String(reflecting: siginfo))")
+    }
+  }
+}
+
+#if !(SWT_TARGET_OS_APPLE && !SWT_NO_LIBDISPATCH)
+/// A mapping of awaited child PIDs to their corresponding Swift continuations.
+private let _childProcessContinuations = Locked<[pid_t: CheckedContinuation<ExitCondition, any Error>]>()
+
+/// A condition variable used to suspend the waiter thread created by
+/// `_createWaitThread()` when there are no child processes to await.
+private let _waitThreadNoChildrenCondition = {
+  let result = UnsafeMutablePointer<pthread_cond_t>.allocate(capacity: 1)
+  _ = pthread_cond_init(result, nil)
+  return UncheckedSendable(rawValue: result)
+}()
+
+#if os(Linux)
+/// Set the name of the current thread.
 ///
-/// This function blocks the calling thread on `waitpid()`. External callers
-/// should use ``wait(for:)`` instead to avoid deadlocks.
-private func _blockAndWait(for pid: pid_t) throws -> ExitCondition {
-  while true {
-    var status: CInt = 0
-    if waitpid(pid, &status, 0) >= 0 {
-      if swt_WIFSIGNALED(status) {
-        return .signal(swt_WTERMSIG(status))
-      } else if swt_WIFEXITED(status) {
-        return .exitCode(swt_WEXITSTATUS(status))
-      } else {
-        // Unreachable: neither signalled nor exited, but waitpid()
-        // and libdispatch indicate that the process has died.
-        throw SystemError(description: "Unexpected waitpid() result \(status). Please file a bug report at https://github.com/apple/swift-testing/issues/new")
+/// This function declaration is provided because `pthread_setname_np()` is
+/// only declared if `_GNU_SOURCE` is set, but setting it causes build errors
+/// due to conflicts with Swift's Glibc module.
+@_extern(c) func pthread_setname_np(_: pthread_t, _: UnsafePointer<CChar>) -> CInt
+#endif
+
+/// The implementation of `_createWaitThread()`, run only once.
+private let _createWaitThreadImpl: Void = {
+  // The body of the thread's run loop.
+  func waitForAnyChild() {
+    // Listen for child process exit events. WNOWAIT means we don't perturb the
+    // state of a terminated (zombie) child process, allowing us to fetch the
+    // continuation (if available) before reaping.
+    var siginfo = siginfo_t()
+    if 0 == waitid(P_ALL, 0, &siginfo, WEXITED | WNOWAIT) {
```

Comment:
Is it possible that `waitid` returns nonzero and sets `errno` to be not `ECHILD`? Since we are calling this method in a `while true` could this loop forever?

Reply:
Possible errors [per POSIX](https://pubs.opengroup.org/onlinepubs/7908799/xsh/waitid.html) are `ECHILD`, `EINTR`, and `EINVAL`. `ECHILD` is handled here; `EINTR` is an interrupt and the correct response is to loop. `EINVAL` is not possible in this circumstance as our inputs are valid. Linux does not AFAIK define any other errors for `waitid()`.